File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/FunctionService.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.table.functions

import org.apache.flink.table.api.ValidationException
import org.apache.flink.table.descriptors._
import org.apache.flink.table.util.Logging

import scala.collection.JavaConversions._
import scala.collection.mutable.ArrayBuffer

/**
  * Service for creating configured instances of [[UserDefinedFunction]] using a
  * [[FunctionDescriptor]].
  */
object FunctionService extends Logging {

  /**
    * Creates a user-defined function with the given properties and the current thread's
    * context class loader.
    *
    * @param descriptor the descriptor that describes a function
    * @return the generated user-defined function
    */
  def createFunction(descriptor: FunctionDescriptor): UserDefinedFunction = {
    createFunction(descriptor, Thread.currentThread().getContextClassLoader)
  }

  /**
    * Creates a user-defined function with the given properties.
    *
    * @param descriptor the descriptor that describes a function
    * @param classLoader the class loader to load the function and its parameter's classes
    * @return the generated user-defined function
    */
  def createFunction(
      descriptor: FunctionDescriptor,
      classLoader: ClassLoader)
    : UserDefinedFunction = {

    val descriptorProperties = new DescriptorProperties(true)
    descriptor.addProperties(descriptorProperties)

    // validate
    new FunctionDescriptorValidator().validate(descriptorProperties)

    // instantiate
    val (instanceClass, instance) = generateInstance[AnyRef](
      HierarchyDescriptorValidator.EMPTY_PREFIX,
      descriptorProperties,
      classLoader)

    if (!classOf[UserDefinedFunction].isAssignableFrom(instanceClass)) {
      throw new ValidationException(
        s"Instantiated class '${instanceClass.getName}' is not a user-defined function.")
    }
    instance.asInstanceOf[UserDefinedFunction]
  }

  /**
   * Recursively generate an instance of a class according the given properties.
   *
   * @param keyPrefix the prefix to fetch properties
   * @param descriptorProperties the descriptor properties that contains the class type information
   * @param classLoader the class loader to load the class
   * @tparam T type fo the generated instance
   * @return an instance of the class
   */
  private def generateInstance[T](
      keyPrefix: String,
      descriptorProperties: DescriptorProperties,
      classLoader: ClassLoader)
    : (Class[T], T) = {

    val instanceClassName = descriptorProperties.getString(
      s"$keyPrefix${ClassInstanceValidator.CLASS}")

    val instanceClass = try {
      Class
        .forName(
          descriptorProperties.getString(s"$keyPrefix${ClassInstanceValidator.CLASS}"),
          true,
          classLoader)
        .asInstanceOf[Class[T]]
    } catch {
      case e: Exception =>
        // only log the cause to have clean error messages
        val msg = s"Could not find class '$instanceClassName' for creating an instance."
        LOG.error(msg, e)
        throw new ValidationException(msg)
    }

    val constructorPrefix = s"$keyPrefix${ClassInstanceValidator.CONSTRUCTOR}"

    val constructorProps = descriptorProperties
      .getVariableIndexedProperties(constructorPrefix, List())

    var i = 0
    val parameterList: ArrayBuffer[(Class[_], Any)] = new ArrayBuffer
    while (i < constructorProps.size()) {
      // nested class instance
      if (constructorProps(i).containsKey(ClassInstanceValidator.CLASS)) {
        parameterList += generateInstance(
          s"$constructorPrefix.$i.",
          descriptorProperties,
          classLoader)
      }
      // literal value
      else {
        val literalValue = LiteralValueValidator
          .getValue(s"$constructorPrefix.$i.", descriptorProperties)
        parameterList += ((literalValue.getClass, literalValue))
      }
      i += 1
    }
    val constructor = try {
      instanceClass.getConstructor(parameterList.map(_._1): _*)
    } catch {
      case e: Exception =>
        // only log the cause to have clean error messages
        val msg = s"Cannot find a public constructor with parameter types " +
          s"'${parameterList.map(_._1.getName).mkString(", ")}' for '$instanceClassName'."
        LOG.error(msg, e)
        throw new ValidationException(msg)
    }

    val instance = try {
      constructor.newInstance(parameterList.map(_._2.asInstanceOf[AnyRef]): _*)
    } catch {
      case e: Exception =>
        // only log the cause to have clean error messages
        val msg = s"Error while creating instance of class '$instanceClassName' " +
          s"with parameter types '${parameterList.map(_._1.getName).mkString(", ")}'."
        LOG.error(msg, e)
        throw new ValidationException(msg)
    }

    (instanceClass, instance)
  }
}

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/FunctionService.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.table.functions;

import org.apache.flink.configuration.Configuration;
import org.apache.flink.configuration.ReadableConfig;
import org.apache.flink.table.api.ValidationException;
import org.apache.flink.table.descriptors.ClassInstanceValidator;
import org.apache.flink.table.descriptors.DescriptorProperties;
import org.apache.flink.table.descriptors.FunctionDescriptor;
import org.apache.flink.table.descriptors.FunctionDescriptorValidator;
import org.apache.flink.table.descriptors.HierarchyDescriptorValidator;
import org.apache.flink.table.descriptors.LiteralValueValidator;
import org.apache.flink.table.descriptors.PythonFunctionValidator;
import org.apache.flink.table.functions.python.utils.PythonFunctionUtils;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.reflect.Constructor;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Service for creating configured instances of {@link UserDefinedFunction} using a
 * {@link FunctionDescriptor}.
 */
public class FunctionService {

	private static final Logger LOG = LoggerFactory.getLogger(FunctionService.class);

	/**
	 * Creates a user-defined function with the given properties and the current thread's
	 * context class loader.
	 *
	 * @param descriptor the descriptor that describes a function
	 * @return the generated user-defined function
	 */
	public static UserDefinedFunction createFunction(FunctionDescriptor descriptor) {
		return createFunction(descriptor, Thread.currentThread().getContextClassLoader());
	}

	/**
	 * Creates a user-defined function with the given properties.
	 *
	 * @param descriptor the descriptor that describes a function
	 * @param classLoader the class loader to load the function and its parameter's classes
	 * @return the generated user-defined function
	 */
	public static UserDefinedFunction createFunction(
			FunctionDescriptor descriptor,
			ClassLoader classLoader) {
		return createFunction(descriptor, classLoader, true);
	}

	/**
	 * Creates a user-defined function with the given properties.
	 *
	 * @param descriptor the descriptor that describes a function
	 * @param classLoader the class loader to load the function and its parameter's classes
	 * @param performValidation whether or not the descriptor should be validated
	 * @return the generated user-defined function
	 */
	public static UserDefinedFunction createFunction(
			FunctionDescriptor descriptor,
			ClassLoader classLoader,
			boolean performValidation) {
		return createFunction(descriptor, classLoader, performValidation, new Configuration());
	}

	/**
	 * Creates a user-defined function with the given properties.
	 *
	 * @param descriptor the descriptor that describes a function
	 * @param classLoader the class loader to load the function and its parameter's classes
	 * @param performValidation whether or not the descriptor should be validated
	 * @param config the table configuration used to create the function
	 * @return the generated user-defined function
	 */
	public static UserDefinedFunction createFunction(
			FunctionDescriptor descriptor,
			ClassLoader classLoader,
			boolean performValidation,
			ReadableConfig config) {

		DescriptorProperties properties = new DescriptorProperties(true);
		properties.putProperties(descriptor.toProperties());

		// validate
		if (performValidation) {
			new FunctionDescriptorValidator().validate(properties);
		}

		Object instance;
		switch (properties.getString(FunctionDescriptorValidator.FROM)) {
			case FunctionDescriptorValidator.FROM_VALUE_CLASS:
				// instantiate
				instance = generateInstance(
					HierarchyDescriptorValidator.EMPTY_PREFIX,
					properties,
					classLoader);
				break;
			case FunctionDescriptorValidator.FROM_VALUE_PYTHON:
				String fullyQualifiedName = properties.getString(PythonFunctionValidator.FULLY_QUALIFIED_NAME);
				instance = PythonFunctionUtils.getPythonFunction(fullyQualifiedName, config);
				break;
			default:
				throw new ValidationException(String.format(
						"Unsupported function descriptor: %s",
						properties.getString(FunctionDescriptorValidator.FROM)));
		}

		if (!UserDefinedFunction.class.isAssignableFrom(instance.getClass())) {
			throw new ValidationException(String.format(
					"Instantiated class '%s' is not a user-defined function.",
					instance.getClass().getName()));
		}
		return (UserDefinedFunction) instance;
	}

	/**
	 * Recursively generate an instance of a class according the given properties.
	 *
	 * @param keyPrefix the prefix to fetch properties
	 * @param descriptorProperties the descriptor properties that contains the class type information
	 * @param classLoader the class loader to load the class
	 * @param <T> type fo the generated instance
	 * @return an instance of the class
	 */
	private static <T> T generateInstance(
			String keyPrefix,
			DescriptorProperties descriptorProperties,
			ClassLoader classLoader) {
		String instanceClassName = descriptorProperties.getString(
				keyPrefix + ClassInstanceValidator.CLASS);

		Class<T> instanceClass;
		try {
			//noinspection unchecked
			instanceClass = (Class<T>) Class.forName(instanceClassName, true, classLoader);
		} catch (Exception e) {
			// only log the cause to have clean error messages
			String msg = String.format(
					"Could not find class '%s' for creating an instance.", instanceClassName);
			LOG.error(msg, e);
			throw new ValidationException(msg);
		}

		String constructorPrefix = keyPrefix + ClassInstanceValidator.CONSTRUCTOR;

		List<Map<String, String>> constructorProps = descriptorProperties
				.getVariableIndexedProperties(constructorPrefix, new ArrayList<>());

		ArrayList<Object> parameterList = new ArrayList<>();
		for (int i = 0; i < constructorProps.size(); i++) {
			String constructorKey = constructorPrefix + "." + i + ".";
			// nested class instance
			if (constructorProps.get(i).containsKey(ClassInstanceValidator.CLASS)) {
				parameterList.add(generateInstance(
						constructorKey,
						descriptorProperties,
						classLoader));
			}
			// literal value
			else {
				Object literalValue = LiteralValueValidator.getValue(
						constructorKey, descriptorProperties);
				parameterList.add(literalValue);
			}
		}

		String parameterNames = parameterList.stream()
				.map(t -> t.getClass().getName())
				.reduce((s1, s2) -> s1 + ", " + s2)
				.orElse("");

		Constructor<T> constructor;
		try {
			constructor = instanceClass.getConstructor(
					parameterList.stream().map(Object::getClass).toArray(Class[]::new));
		} catch (Exception e) {
			// only log the cause to have clean error messages
			String msg = String.format(
					"Cannot find a public constructor with parameter types '%s' for '%s'.",
					parameterNames,
					instanceClassName);
			LOG.error(msg, e);
			throw new ValidationException(msg);
		}

		try {
			return constructor.newInstance(parameterList.toArray());
		} catch (Exception e) {
			// only log the cause to have clean error messages
			String msg = String.format(
					"Error while creating instance of class '%s' with parameter types '%s'.",
					instanceClassName,
					parameterNames);
			LOG.error(msg, e);
			throw new ValidationException(msg);
		}
	}
}
